### PR TITLE
Add group & ungroup

### PR DESCRIPTION
**Pull request by binh-vu**

### CODE REVIEW

File: www/src/models/dedupMineralSite/DedupMineralSiteStore.ts [hunk 1]
```diff
@@ -16,13 +17,44 @@ export class DedupMineralSiteStore extends RStore<string, DedupMineralSite> {
...
+
+  /**
+   * Replace given dedup sites with new sites
+   *
+   * @param prevIds previous sites to delete
+   * @param newIds new sites to add
+   */
+  async replaceSites(prevIds: InternalID[], newIds: InternalID[], commodity:Object): Promise<void> {
+    console.log("prevIds",prevIds)
+    console.log("newIds",newIds)
```
Comment (binh-vu) [inline www/src/models/dedupMineralSite/DedupMineralSiteStore.ts:53]: remove it

File: www/src/models/dedupMineralSite/DedupMineralSiteStore.ts [hunk 2]
```diff
@@ -16,13 +17,44 @@ export class DedupMineralSiteStore extends RStore<string, DedupMineralSite> {
...
+    this.deleteByIds(prevIds);
+    const newSites = await this.fetchByIds(newIds, true, {commodity});
```
Comment (binh-vu) [inline www/src/models/dedupMineralSite/DedupMineralSiteStore.ts:55]: remove assignment

File: www/src/pages/home/DedupMineralSiteTable.tsx [hunk 3]
```diff
@@ -31,7 +33,7 @@ const columns = [
         </Typography.Link>
       );
     },
-    sorter: (a: DedupMineralSite, b: DedupMineralSite) => a.name.localeCompare(b.name),
+      sorter: (a: DedupMineralSite, b: DedupMineralSite) => a.name.localeCompare(b.name),
```
Comment (binh-vu) [inline www/src/pages/home/DedupMineralSiteTable.tsx:36]: format

File: www/src/pages/home/DedupMineralSiteTable.tsx [hunk 4]
```diff
@@ -58,6 +60,11 @@ const columns = [
       }
       return "-";
     },
+    sorter: (a: DedupMineralSite, b: DedupMineralSite) => {
+      const locA = a.location ? `${a.location.lat},${a.location.lon}` : "";
+      const locB = b.location ? `${b.location.lat},${b.location.lon}` : "";
```
Comment (binh-vu) [inline www/src/pages/home/DedupMineralSiteTable.tsx:65]: use toFixed(3)

File: www/src/pages/home/DedupMineralSiteTable.tsx [hunk 5]
```diff
@@ -75,6 +82,11 @@ const columns = [
         </Space>
       );
     },
+    sorter: (a: DedupMineralSite, b: DedupMineralSite) => {
+      const countryA = a.location?.country.join(",") || "";
+      const countryB = b.location?.country.join(",") || "";
```
Comment (binh-vu) [inline www/src/pages/home/DedupMineralSiteTable.tsx:87]: sort by name, country is an object

File: www/src/pages/home/DedupMineralSiteTable.tsx [hunk 6]
```diff
@@ -92,6 +104,11 @@ const columns = [
         </Space>
       );
     },
+    sorter: (a: DedupMineralSite, b: DedupMineralSite) => {
+      const stateA = a.location?.stateOrProvince.join(",") || "";
+      const stateB = b.location?.stateOrProvince.join(",") || "";
+      return stateA.localeCompare(stateB);
```
Comment (binh-vu) [inline www/src/pages/home/DedupMineralSiteTable.tsx:110]: same of country

File: www/src/pages/home/DedupMineralSiteTable.tsx [hunk 7]
```diff
@@ -145,17 +172,100 @@ const columns = [
 export const DedupMineralSiteTable: React.FC<DedupMineralSiteTableProps> = observer(({ commodity }) => {
   const { dedupMineralSiteStore, commodityStore } = useStores();
   const [editingDedupSite, setEditingDedupSite] = useState<string | undefined>(undefined);
+  const [movedRows, setMovedRows] = useState<DedupMineralSite[]>([]);
+  const [currentRows, setCurrentRows] = useState<DedupMineralSite[]>([]);
+  const [groupSuccess, setGroupSuccess] = useState(false);
+  const [showStickyDiv, setShowStickyDiv] = useState(true);
```
Comment (binh-vu) [inline www/src/pages/home/DedupMineralSiteTable.tsx:178]: remove sticky

File: www/src/pages/home/DedupMineralSiteTable.tsx [hunk 8]
```diff
@@ -145,17 +172,100 @@ const columns = [
 export const DedupMineralSiteTable: React.FC<DedupMineralSiteTableProps> = observer(({ commodity }) => {
   const { dedupMineralSiteStore, commodityStore } = useStores();
   const [editingDedupSite, setEditingDedupSite] = useState<string | undefined>(undefined);
+  const [movedRows, setMovedRows] = useState<DedupMineralSite[]>([]);
```
Comment (binh-vu) [inline www/src/pages/home/DedupMineralSiteTable.tsx:175]: rename to selectedDedupSites

File: www/src/pages/home/DedupMineralSiteTable.tsx [hunk 9]
```diff
@@ -176,26 +286,88 @@ export const DedupMineralSiteTable: React.FC<DedupMineralSiteTableProps> = obser
...
+  }}
+>
+  Ungroup
+</Button>
+
       </Space>
     );
   };
 
+  columns[0].render = (_: any, site: DedupMineralSite) => <Checkbox onChange={(e) => handleRowMove(site, e.target.checked)} checked={movedRows.some((row) => row.id === site.id)} />;
```
Comment (binh-vu) [inline www/src/pages/home/DedupMineralSiteTable.tsx:310]: change to using a set

File: www/src/pages/home/DedupMineralSiteTable.tsx [hunk 10]
```diff
@@ -145,17 +172,100 @@ const columns = [
 export const DedupMineralSiteTable: React.FC<DedupMineralSiteTableProps> = observer(({ commodity }) => {
   const { dedupMineralSiteStore, commodityStore } = useStores();
   const [editingDedupSite, setEditingDedupSite] = useState<string | undefined>(undefined);
+  const [movedRows, setMovedRows] = useState<DedupMineralSite[]>([]);
+  const [currentRows, setCurrentRows] = useState<DedupMineralSite[]>([]);
```
Comment (binh-vu) [inline www/src/pages/home/DedupMineralSiteTable.tsx:176]: Replaced selectedDedupSites

File: www/src/pages/home/DedupMineralSiteTable.tsx [hunk 11]
```diff
@@ -145,17 +172,92 @@ const columns = [
...
-    if (commodity !== undefined) {
-      dedupMineralSiteStore.fetchByCommodity(commodity);
-    }
-  }, [commodity]);
+    const fetchData = async () => {
+      if (commodity) {
+        try {
+          const result = await dedupMineralSiteStore.fetchByCommodity(commodity);
+        } catch (error) {
+          console.error("Error fetching data:", error);
```
Comment (binh-vu) [inline www/src/pages/home/DedupMineralSiteTable.tsx:186]: remove console

File: www/src/pages/home/DedupMineralSiteTable.tsx [hunk 12]
```diff
@@ -145,17 +172,92 @@ const columns = [
...
+      setSelectedDedupSiteIds(selectedDedupSiteIds.union(new Set([site.id])));
+    } else {
+      setSelectedDedupSiteIds(selectedDedupSiteIds.difference(new Set([site.id])));
+    }
+  };
+
+  const handleGroup = async () => {
+    try {
+      const prevIds = Array.from(selectedDedupSiteIds);
+      console.log("OLD Ids", prevIds);
```
Comment (binh-vu) [inline www/src/pages/home/DedupMineralSiteTable.tsx:211]: delete

File: www/src/pages/home/DedupMineralSiteTable.tsx [hunk 13]
```diff
@@ -145,17 +172,92 @@ const columns = [
...
+
+      if (prevIds.length === 0) {
+        alert("No site IDs found for grouping. Please add some rows.");
+        return;
+      }
```
Comment (binh-vu) [inline www/src/pages/home/DedupMineralSiteTable.tsx:216]: check if this condition is needed and remove it if not

File: www/src/pages/home/DedupMineralSiteTable.tsx [hunk 14]
```diff
@@ -145,17 +172,92 @@ const columns = [
...
+
+      const allSiteIds = Array.from(selectedDedupSiteIds).flatMap((dedupSiteId) => dedupMineralSiteStore.get(dedupSiteId)!.sites.map((siteUri) => DedupMineralSite.getId(siteUri)));
+
+      const sameAsPayload = [
+        {
+          sites: allSiteIds,
+        },
+      ];
+
+      console.log("Grouping Payload:", sameAsPayload);
```
Comment (binh-vu) [inline www/src/pages/home/DedupMineralSiteTable.tsx:226]: remove

File: www/src/pages/home/DedupMineralSiteTable.tsx [hunk 15]
```diff
@@ -145,17 +172,92 @@ const columns = [
...
+
+      try {
+        const sameAsResponse = await axios.post("/api/v1/same-as", sameAsPayload, {
+          headers: {
+            "Content-Type": "application/json",
+          },
+          withCredentials: true,
+        });
```
Comment (binh-vu) [inline www/src/pages/home/DedupMineralSiteTable.tsx:234]: move this to a dedup store

File: www/src/pages/home/DedupMineralSiteTable.tsx [hunk 16]
```diff
@@ -145,17 +172,92 @@ const columns = [
...
+        console.log("Group API Response:", sameAsResponse.data);
+
+        const newIds = sameAsResponse.data.map((item: any) => item.id);
+
+        console.log("New Site IDs:", newIds);
+
+        if (commodity && commodity.id) {
+          const commodityId = commodity.id;
+          await dedupMineralSiteStore.replaceSites(prevIds, newIds, commodityId);
+          setGroupSuccess(true);
```
Comment (binh-vu) [inline www/src/pages/home/DedupMineralSiteTable.tsx:244]: change groupsuccess to a temporal message

File: www/src/pages/home/DedupMineralSiteTable.tsx [hunk 17]
```diff
@@ -145,17 +172,92 @@ const columns = [
...
+          setSelectedDedupSiteIds(new Set());
+        } else {
+          console.error("commodity is undefined or does not have an id");
+        }
+        console.log("Sites replaced successfully.");
+      } catch (error) {
+        console.error("Error during /same-as API call:", error);
+        alert("Group operation failed. Please check the console for details.");
+      }
```
Comment (binh-vu) [inline www/src/pages/home/DedupMineralSiteTable.tsx:254]: no need to handle this error

File: www/src/pages/home/DedupMineralSiteTable.tsx [hunk 18]
```diff
@@ -176,26 +278,78 @@ export const DedupMineralSiteTable: React.FC<DedupMineralSiteTableProps> = obser
...
-    />
+    <>
+      {groupSuccess && <Alert message="Grouping successful!" type="success" showIcon closable afterClose={() => setGroupSuccess(false)} />}
+      {showStickyDiv && selectedDedupSiteIds.size > 0 && (
+        <div style={{ position: "sticky", top: 0, zIndex: 1000, background: "#fff", marginTop: "16px" }}>
+          <Typography.Title level={4}>Moved Rows</Typography.Title>
+          <Button type="primary" onClick={handleGroup}>
+            Group
+          </Button>
+          <Table<DedupMineralSite>
```
Comment (binh-vu) [inline www/src/pages/home/DedupMineralSiteTable.tsx:311]: remove pagination

File: www/src/pages/home/DedupMineralSiteTable.tsx [hunk 19]
```diff
@@ -176,26 +278,73 @@ export const DedupMineralSiteTable: React.FC<DedupMineralSiteTableProps> = obser
...
-        expandedRowRender: (site) => <EditDedupMineralSite commodity={commodity!} dedupSite={site} />,
-        showExpandColumn: false,
-        expandedRowKeys: editingDedupSite === undefined ? [] : [editingDedupSite],
-      }}
-    />
+    <>
+      {groupSuccess && <Alert message="Grouping successful!" type="success" showIcon closable afterClose={() => setGroupSuccess(false)} />}
+      {showStickyDiv && selectedDedupSiteIds.size > 0 && (
+        <div style={{ position: "sticky", top: 0, zIndex: 1000, background: "#fff", marginTop: "16px" }}>
+          <Button type="primary" onClick={handleGroup}>
```
Comment (binh-vu) [inline www/src/pages/home/DedupMineralSiteTable.tsx:308]: add space

File: www/src/pages/home/DedupMineralSiteTable.tsx (hunk 5 above)
Comment (NithyashreeGS30) [reply www/src/pages/home/DedupMineralSiteTable.tsx:86]: it says name does not exist

File: www/src/pages/home/DedupMineralSiteTable.tsx (hunk 6 above)
Comment (NithyashreeGS30) [reply www/src/pages/home/DedupMineralSiteTable.tsx:109]: same for this

File: www/src/models/dedupMineralSite/DedupMineralSiteStore.ts [hunk 20]
```diff
@@ -109,7 +139,65 @@ export class DedupMineralSiteStore extends RStore<string, DedupMineralSite> {
...
+        headers: {
+          "Content-Type": "application/json",
+        },
+        withCredentials: true,
+      });
+  
+      return response.data;
+    } catch (error) {
+      console.error("Error during groupMineralSites:", error);
+      throw error;
```
Comment (binh-vu) [inline www/src/models/dedupMineralSite/DedupMineralSiteStore.ts:154]: no need for try catch

File: www/src/models/dedupMineralSite/DedupMineralSiteStore.ts [hunk 21]
```diff
@@ -109,7 +139,65 @@ export class DedupMineralSiteStore extends RStore<string, DedupMineralSite> {
     });
   }
 
+  async CreateOnegroup(sameAsPayload: any): Promise<any> {
```
Comment (binh-vu) [inline www/src/models/dedupMineralSite/DedupMineralSiteStore.ts:142]: better name & update arguments

File: www/src/models/dedupMineralSite/DedupMineralSiteStore.ts [hunk 22]
```diff
@@ -109,7 +139,65 @@ export class DedupMineralSiteStore extends RStore<string, DedupMineralSite> {
...
+          "Content-Type": "application/json",
+        },
+        withCredentials: true,
+      });
+  
+      return response.data; 
+    } catch (error) {
+      console.error("Error during groupMineralSites:", error);
+      throw error; 
+    }
```
Comment (binh-vu) [inline www/src/models/dedupMineralSite/DedupMineralSiteStore.ts:170]: silimiar

File: www/src/models/dedupMineralSite/DedupMineralSiteStore.ts [hunk 23]
```diff
@@ -109,7 +139,65 @@ export class DedupMineralSiteStore extends RStore<string, DedupMineralSite> {
...
+        withCredentials: true,
+      });
+  
+      return response.data;
+    } catch (error) {
+      console.error("Error during groupMineralSites:", error);
+      throw error; 
+    }
+  }
+  async CreateKgroups(payload: any): Promise<any> {
```
Comment (binh-vu) [inline www/src/models/dedupMineralSite/DedupMineralSiteStore.ts:157]: arguments need to have type

File: www/src/models/dedupMineralSite/DedupMineralSiteStore.ts [hunk 24]
```diff
@@ -99,17 +129,68 @@ export class DedupMineralSiteStore extends RStore<string, DedupMineralSite> {
...
+        },
+        withCredentials: true,
+      });
+
+      return response.data;
+    } catch (error) {
+      console.error("Error during groupMineralSites:", error);
+      throw error;
+    }
+  }
```
Comment (binh-vu) [inline www/src/models/dedupMineralSite/DedupMineralSiteStore.ts:171]: remove and other code use updateSameAsGroup

File: www/src/pages/home/DedupMineralSiteTable.tsx [hunk 25]
```diff
@@ -1,201 +1,295 @@
...
-      if (site.location === undefined) {
-        return "-";
-      }
+export const DedupMineralSiteTable: React.FC<DedupMineralSiteTableProps> = observer(({ commodity }) => {
+  const { dedupMineralSiteStore, countryStore, stateOrProvinceStore } = useStores();
+  const [editingDedupSite, setEditingDedupSite] = useState<string | undefined>(undefined);
+  const [selectedDedupSiteIds, setSelectedDedupSiteIds] = useState<Set<string>>(new Set());
+  const [messageApi, contextHolder] = message.useMessage();
+  const [showDiv, setShowDiv] = useState(true);
+  const [ungroupingSite, setUngroupingSite] = useState<string | undefined>(undefined);
```
Comment (binh-vu) [inline www/src/pages/home/DedupMineralSiteTable.tsx:24]: remove redundant variables

File: www/src/pages/home/DedupMineralSiteTable.tsx [hunk 26]
```diff
@@ -1,201 +1,295 @@
...
-    if (commodity !== undefined) {
-      dedupMineralSiteStore.fetchByCommodity(commodity);
-    }
-  }, [commodity]);
+    const fetchData = async () => {
+      if (commodity) {
+        try {
+          const result = await dedupMineralSiteStore.fetchByCommodity(commodity);
+        } catch (error) {
+          console.error("Error fetching data:", error);
```
Comment (binh-vu) [inline www/src/pages/home/DedupMineralSiteTable.tsx:208]: remove this

File: www/src/pages/home/DedupMineralSiteTable.tsx [hunk 27]
```diff
@@ -1,201 +1,295 @@
...
+    };
+    fetchData();
+  }, [commodity, dedupMineralSiteStore]);
 
   if (dedupMineralSiteStore.state.value === "error") {
     return <Alert message="Error" description="An error occurred while querying dedup mineral sites. Please try again later." type="error" showIcon />;
   }
 
+  const selectDedupSite = (site: DedupMineralSite, selectOrNot: boolean) => {
+    setShowDiv(true);
```
Comment (binh-vu) [inline www/src/pages/home/DedupMineralSiteTable.tsx:220]: remove this

File: www/src/pages/home/DedupMineralSiteTable.tsx [hunk 28]
```diff
@@ -1,201 +1,295 @@
...
-      loading={isLoading ? { size: "large" } : false}
-      expandable={{
-        expandedRowRender: (site) => <EditDedupMineralSite commodity={commodity!} dedupSite={site} />,
-        showExpandColumn: false,
-        expandedRowKeys: editingDedupSite === undefined ? [] : [editingDedupSite],
-      }}
-    />
+    <>
+      {showDiv && selectedDedupSiteIds.size > 0 && (
+        <div style={{ top: 0, zIndex: 1000, background: "#fff", marginTop: "16px" }}>
```
Comment (binh-vu) [inline www/src/pages/home/DedupMineralSiteTable.tsx:254]: no longer need these zindex. use normal div, and space after button

File: www/src/pages/home/DedupMineralSiteTable.tsx [hunk 29]
```diff
@@ -1,201 +1,295 @@
...
+        loading={isLoading ? { size: "large" } : false}
+        expandable={{
+          expandedRowRender: (site) => {
+            if (editingDedupSite === site.id) {
+              return <EditDedupMineralSite commodity={commodity!} dedupSite={site} />;
+            }
+            return null;
+          },
+          showExpandColumn: false,
+          expandedRowKeys: [...(editingDedupSite ? [editingDedupSite] : []), ...(ungroupingSite ? [ungroupingSite] : [])],
```
Comment (binh-vu) [inline www/src/pages/home/DedupMineralSiteTable.tsx:290]: ungroupingSite remove?

File: www/src/pages/home/editDedupSite/EditDedupMineralSite.tsx [hunk 30]
```diff
@@ -36,11 +36,114 @@ interface EditDedupMineralSiteProps {
...
+    const isLoading = mineralSiteStore.state.value === "updating" || fetchedSites.length !== dedupSite.sites.length;
+
+    const handleCreateOne = async () => {
+      const selectedSiteIds = Array.from(selectedRows);
+
+      const allSiteIds = sites.map((site) => site.id);
+
+      const unselectedSiteIds = allSiteIds.filter((id) => !selectedRows.has(id));
+
+      const sameAsPayload =
```
Comment (binh-vu) [inline www/src/pages/home/editDedupSite/EditDedupMineralSite.tsx:57]: better name

File: www/src/pages/home/editDedupSite/EditDedupMineralSite.tsx [hunk 31]
```diff
@@ -36,11 +36,114 @@ interface EditDedupMineralSiteProps {
...
+
+      if (commodity && commodity.id) {
+        const commodityId = commodity.id;
+        await dedupMineralSiteStore.replaceSites([dedupSite.id], newIds, commodityId);
+        message.success("Operation was successful!");
+
+      }
+    };
+
+    const handleKGroup = async () => {
```
Comment (binh-vu) [inline www/src/pages/home/editDedupSite/EditDedupMineralSite.tsx:76]: rename the function to match with text description

File: www/src/pages/home/editDedupSite/EditDedupMineralSite.tsx [hunk 32]
```diff
@@ -36,11 +36,114 @@ interface EditDedupMineralSiteProps {
...
+      const createPayload = [...selectedPayload, ...unselectedPayload];
+
+      const response = await dedupMineralSiteStore.createKgroups(createPayload);
+
+      const newIds = response.data.map((item: any) => item.id);
+
+      if (commodity && commodity.id) {
+        const commodityId = commodity.id;
+        await dedupMineralSiteStore.replaceSites([dedupSite.id], newIds, commodityId);
+        message.success("Operation was successful!");
```
Comment (binh-vu) [inline www/src/pages/home/editDedupSite/EditDedupMineralSite.tsx:94]: better name for this message

**Pull request merged**